File: src/main/scala/uk/ac/warwick/courses/validators/SpelAssertValidator.scala
package uk.ac.warwick.courses.validators;

import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext
import org.springframework.expression.ExpressionParser
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.expression.Expression
import org.springframework.expression.spel.standard.SpelExpressionParser

/**
 * Hibernate validators has @ScriptAssert which lets you make scripted validation checks.
 * Spring has SpEL which is a good script language, but it's not in the Java Scripting JSR
 * so use this thing we made.
 */
class SpelAssertValidator extends ConstraintValidator[SpelAssert, Object] {
  
	@Autowired var parser:ExpressionParser =_
	var expression: Expression =_
  
	override def initialize(annotation:SpelAssert) {
	  expression = parser.parseExpression(annotation.value())
	}
	
	override def isValid(value:Object, ctx:ConstraintValidatorContext) = value match {
	  case value:Object => expression.getValue(value, classOf[Boolean])
	  case _ => true // nothing to run against
	}
}

File: src/main/java/uk/ac/warwick/courses/validators/SpelAssertValidator.java
package uk.ac.warwick.courses.validators;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;
import javax.validation.ConstraintValidatorContext.ConstraintViolationBuilder;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.expression.Expression;
import org.springframework.expression.ExpressionParser;

public class SpelAssertValidator implements ConstraintValidator<SpelAssert, Object> {

	@Autowired private ExpressionParser parser;
	private Expression expression;
	private String message;
	private String targetProperty;
	
	@Override
	public void initialize(SpelAssert annotation) {
		message = annotation.message();
		targetProperty = annotation.targetProperty();
		expression = parser.parseExpression(annotation.value());
	}

	@Override
	public boolean isValid(Object object, ConstraintValidatorContext ctx) {
		if (object == null) return true;
		else {
			boolean valid = expression.getValue(object, Boolean.class);
			if (!valid) {
				if (!"".equals(targetProperty)) {
					ctx.disableDefaultConstraintViolation();
					ConstraintViolationBuilder builder = ctx.buildConstraintViolationWithTemplate(message);
					builder.addNode(targetProperty).addConstraintViolation();
				}
			}
			return valid;
		}
	}

}
